File: embeddinghub/embeddingstore/serializer.cc
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

#include "serializer.h"

#include "embeddingstore/embedding_store.grpc.pb.h"

namespace featureform {

namespace embedding {

std::vector<float> ProtoSerializer::deserialize(
    const std::string& serialized) const {
  auto proto = proto::Embedding();
  proto.ParseFromString(serialized);
  auto vals = proto.values();
  return std::vector<float>(vals.begin(), vals.end());
}

std::string ProtoSerializer::serialize(std::vector<float> vals) const {
  auto proto = proto::Embedding();
  *proto.mutable_values() = {vals.begin(), vals.end()};
  std::string serialized;
  proto.SerializeToString(&serialized);
  return serialized;
}

}  // namespace embedding
}  // namespace featureform

File: provider/serialization/serializer.go
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright 2024 FeatureForm Inc.
//

package serialization

import (
	vt "github.com/featureform/provider/types"
	"strconv"
)

// SerializeVersion is used to specify what method of serializing and deserializing values
// that we're using.
type SerializeVersion int

func (v SerializeVersion) String() string {
	return strconv.Itoa(int(v))
}

// Serializer provides methods to serialize and deserialize values.
type Serializer[T any] interface {
	Version() SerializeVersion
	Serialize(t vt.ValueType, value any) (T, error)
	Deserialize(t vt.ValueType, value T) (any, error)
}
